### PR TITLE
Partial waypoint uploading support and bug fix in waypoint uploading

### PR DESCRIPTION
**Pull request by JMare**
This PR adds support for partial waypoint uploading with a new service including support in mavwp cli script.

Also fixed a bug we uncovered when uploading missions to APM.

When state is TXWP and mission item N is received by FCU, FCU sends request for N+1.
If that request packet gets dropped, mavros will resend mission_item N.
At that point FCU will be expecting N+1 and mavros will be sending N.
FCU responds by sending a mission_ack of type "invalid_sequence".
This causes mavros to abort the transfer and go_idle() even though the FCU is still requesting N+1.

This PR makes mavros stay in TXWP state when that particular mission_ack is received, so the transfer will complete when the autopilot resends the request for N+1.

### CODE REVIEW

**Comment by JMare** [comment]
@TSC21 Sure I can make a video tomorrow of partial waypoint uploading to APM SITL.

One of the CI checks failed, do you know what tests it runs? I ran catkin run_tests and it seemed ok.

File: mavros/scripts/mavwp [hunk 1]
```diff
@@ -187,6 +207,12 @@ def main():
     load_args.add_argument('-p', '--preserve-home', action='store_true', help="Preserve home location (WP 0, APM only)")
     load_args.add_argument('file', type=argparse.FileType('rb'), help="input file (QGC/MP format)")
 
+    load_partial_args = subarg.add_parser('load_partial', help="partially load waypoints from file")
+    load_partial_args.set_defaults(func=do_load_partial)
+    load_partial_args.add_argument('file', type=argparse.FileType('rb'), help="input file (QGC/MP format)")
+    load_partial_args.add_argument('start_index', type=int, help="waypoint start index")
+    load_partial_args.add_argument('end_index', type=int, help="waypoint end index")
```
Comment (vooon) [inline mavros/scripts/mavwp]: I think that indexes better set via options, not positional args. Start index should be `required=True`; End - by default use last entry.

File: mavros/scripts/mavwp [hunk 2]
```diff
@@ -187,6 +207,12 @@ def main():
     load_args.add_argument('-p', '--preserve-home', action='store_true', help="Preserve home location (WP 0, APM only)")
     load_args.add_argument('file', type=argparse.FileType('rb'), help="input file (QGC/MP format)")
 
+    load_partial_args = subarg.add_parser('load_partial', help="partially load waypoints from file")
```
Comment (vooon) [inline mavros/scripts/mavwp]: Name it `load-partial`.

**Comment by TSC21** [comment]
> One of the CI checks failed, do you know what tests it runs? I ran catkin run_tests and it seemed ok.

Nothing to worry about the CI. This is a Jenkins related problem ;)

**Comment by TSC21** [comment]
@JMare any updates?

**Comment by TSC21** [comment]
Jenkins test this please.

**Comment by JMare** [comment]
@TSC21 We implemented the changes @vooon suggested.
In doing so we realised that our implementation was leaving send_waypoints in a bad state so we are trying to fix that.

Our latest work is in https://github.com/MonashUAS/mavros/tree/partial_tmp
Should have it done in a couple of days.

**Comment by TSC21** [comment]
@JMare perfect thank you! Do you have a way of testing the same with an autopilot using PX4 Firmware?

**Comment by JMare** [comment]
@TSC21 I have never tried but PX4 SITL looks easy enough since I am familiar with gazebo.
I'll give it a go tommorow.

**Comment by TSC21** [comment]
> @TSC21 I have never tried but PX4 SITL looks easy enough since I am familiar with gazebo.
I'll give it a go tommorow.

Great thank you!

**Comment by JMare** [comment]
@JamesStewy Managed to iron out the bugs.

Having a hard time with PX4 Sitl, going to just flash a Pixhawk to test over the next few days.

**Comment by TSC21** [comment]
What problems are you facing?

**Comment by JMare** [comment]
@TSC21 I had build errors on the gazebo plugins.
I am using a non standard build of gazebo though.

Today I tested with a Pixhawk running PX4 Firmware and partial way point updating does not work.

Doing some searching and looking through https://github.com/PX4/Firmware/blob/master/src/modules/mavlink/mavlink_mission.cpp

It does not look like PX4 Firmware implements partial waypoint updating at all. 
This feature may have to be APM specific.

**Comment by TSC21** [comment]
Ok just be sure it does not break the normal waypoiny feed to the PX4 Firmware then. Please also document it does not currently work for the PX4 Firmware for lack of support on the Firmware side. Thanks

**Comment by JMare** [comment]
Yes I checked and was still able to waypoint load/clear/pull as normal so it doesn't break that.

Should we make the feature enabled only for APM FCU as specified in launch files?

**Comment by TSC21** [comment]
Yes please. If you may, please open an issue on the PX4 firmware as a feature request, so to be tracked.

File: mavros/src/plugins/waypoint.cpp [hunk 3]
```diff
@@ -147,6 +147,7 @@ class WaypointPlugin : public plugin::PluginBase {
 		wp_state = WP::IDLE;
 
 		wp_nh.param("pull_after_gcs", do_pull_after_gcs, false);
+        wp_nh.param("enable_partial_push",enable_partial_push,false);
```
Comment (TSC21) [inline mavros/src/plugins/waypoint.cpp]: use tab for indentation

File: mavros/src/plugins/waypoint.cpp [hunk 4]
```diff
@@ -215,6 +216,8 @@ class WaypointPlugin : public plugin::PluginBase {
 	ros::Timer wp_timer;
 	ros::Timer shedule_timer;
 	bool do_pull_after_gcs;
+    bool enable_partial_push;
```
Comment (TSC21) [inline mavros/src/plugins/waypoint.cpp]: use tab for indentation

File: mavros/src/plugins/waypoint.cpp [hunk 5]
```diff
@@ -736,6 +739,14 @@ class WaypointPlugin : public plugin::PluginBase {
...
 		{
 			//Partial Waypoint update
+
+            if(!enable_partial_push)
+            {
+                ROS_WARN_NAMED("wp","Partial Push not enabled. (Only supported on APM)");
+                res.success = false;
+                res.wp_transfered = 0;
+                return true;
+            }
```
Comment (TSC21) [inline mavros/src/plugins/waypoint.cpp]: don't use spaces. Use tab for indentation (equivalent to 8 spaces).

Comment (TSC21) [reply mavros/src/plugins/waypoint.cpp]: Or just run the uncrustify tool on the file (`uncrustify -c ${ROS_WORKSPACE}/mavros/mavros/tools/uncrustify-cpp.cfg --replace --no-backup <path/to/file.ext>`)

**Comment by TSC21** [comment]
Two last things:
1. Solve the conflict on the .cpp file;
2. Apply the Doxygen-like commenting for documentation on the full file, as in https://github.com/mavlink/mavros/blob/master/mavros/src/plugins/imu.cpp, if you may. (It will help me out with https://github.com/mavlink/mavros/issues/802). Thanks

**Comment by vooon** [comment]
Hmm, unfortunately i do not see partial mission in `MAV_PROTOCOL_CAPABILITY`.
But you may assume that if FCU is APM then that feature is supported.

File: mavros/scripts/mavwp [hunk 6]
```diff
@@ -118,17 +117,28 @@ def do_load(args):
...
-        _load_call(wps)
+    if args.start_index > 0:
+        # Push partial
+        if args.start_index == args.end_index:
+            args.end_index += 1
+
+        if args.end_index == 0:
+            _load_call(args.start_index, wps[args.start_index:])
+        else:
+            _load_call(args.start_index, wps[args.start_index:args.end_index])
```
Comment (vooon) [inline mavros/scripts/mavwp]:
```python
end_index = args.end_index or len(wps)
_load_call(args.start_index, wps[args.start_index:end_index])
```

File: mavros/scripts/mavwp [hunk 7]
```diff
@@ -185,6 +195,8 @@ def main():
     load_args = subarg.add_parser('load', help="load waypoints from file")
     load_args.set_defaults(func=do_load)
     load_args.add_argument('-p', '--preserve-home', action='store_true', help="Preserve home location (WP 0, APM only)")
+    load_args.add_argument('-s', '--start-index', type=int, default=0, help="waypoint start index for partial updating")
+    load_args.add_argument('-e', '--end-index', type=int, default=0, help="waypoint end index for partial updating")
```
Comment (vooon) [inline mavros/scripts/mavwp]: Add to help "(default: last element in waypoint list)"

File: mavros/src/plugins/waypoint.cpp [hunk 8]
```diff
@@ -147,6 +147,7 @@ class WaypointPlugin : public plugin::PluginBase {
 		wp_state = WP::IDLE;
 
 		wp_nh.param("pull_after_gcs", do_pull_after_gcs, false);
+		wp_nh.param("enable_partial_push",enable_partial_push,false);
```
Comment (vooon) [inline mavros/src/plugins/waypoint.cpp]: Better:

```
//! @note currently (2017) only APM support partial mission update
wp_nh.param("enable_partial_push", enable_partial_push, m_uas->is_ardupilotmega());
```

**Comment by TSC21** [comment]
> But you may assume that if FCU is APM then that feature is supported.

Yeah I suppose that instead of a parameter we can check for the FCU type by checking `UAS::MAV_AUTOPILOT`.

File: mavros/src/plugins/waypoint.cpp [hunk 9]
```diff
@@ -649,18 +663,30 @@ class WaypointPlugin : public plugin::PluginBase {
...
+	void mission_write_partial_list(uint16_t start_index, uint16_t end_index)
+	{
+		ROS_DEBUG_NAMED("wp", "WP:m: write partial list %u - %u", start_index, end_index);
+
+		mavlink::common::msg::MISSION_WRITE_PARTIAL_LIST mwpl {};
+		mwpl.start_index = start_index;
+		mwpl.end_index = end_index;
+
+		m_uas->msg_set_target(mwpl);
+		UAS_FCU(m_uas)->send_message_ignore_drop(mwpl);
```
Comment (vooon) [inline mavros/src/plugins/waypoint.cpp]: `msg_set_target()` should be in msg fill group of lines.

File: mavros/src/plugins/waypoint.cpp [hunk 10]
```diff
@@ -702,39 +728,86 @@ class WaypointPlugin : public plugin::PluginBase {
...
 	{
 		unique_lock lock(mutex);
 
 		if (wp_state != WP::IDLE)
 			// Wrong initial state, other operation in progress?
 			return false;
 
-		wp_state = WP::TXLIST;
+		if (req.start_index)
+		{
```
Comment (vooon) [inline mavros/src/plugins/waypoint.cpp]: Style:
```
if () {
}
else if () {
}
else {
}
```

Unfortunately uncrustify do not fix that.

File: mavros/src/plugins/waypoint.cpp [hunk 11]
```diff
@@ -702,39 +728,86 @@ class WaypointPlugin : public plugin::PluginBase {
...
+				return true;
+			}
+
+			if (waypoints.size() < req.start_index + req.waypoints.size())
+			{
+				ROS_WARN_NAMED("wp","WP: Partial push out of range rejected.");
+				res.success = false;
+				res.wp_transfered = 0;
+				return true;
+			}
```
Comment (vooon) [inline mavros/src/plugins/waypoint.cpp]: Partial update does not support extension of existing mission?

File: mavros_msgs/srv/WaypointPush.srv [hunk 12]
```diff
@@ -1,7 +1,8 @@
 # Send waypoints to device
-#
+# Start index will define a partial waypoint update
 # Returns success status and transfered count
```
Comment (vooon) [inline mavros_msgs/srv/WaypointPush.srv]: Better edit that style:
```
# Send waypoints to device
#
#  :start_index: will define a partial waypoint update. Set to 0 for full update.
#
# Returns success status and transferred count
```

File: mavros/src/plugins/waypoint.cpp (hunk 8 above)
Comment (vooon) [reply mavros/src/plugins/waypoint.cpp]: Oh, no, that would not work.

here that might be defaulted to `false` but in `connection_cb()` need to check:

```
if (!wp_nh.getParam(...)) {
// no override param - detect by FCU
enable = m_uas->is_ardupilotmega();
}
```

**Comment by vooon** [comment]
I think that parameter is fine, it might be used as override.

File: mavros/src/plugins/waypoint.cpp (hunk 8 above)
Comment (JMare) [reply mavros/src/plugins/waypoint.cpp]: Yes this is better. Then i will remove it from the APM config yaml.

File: mavros/src/plugins/waypoint.cpp (hunk 11 above)
Comment (JMare) [reply mavros/src/plugins/waypoint.cpp]: No, it doesn't the waypoint count cant change

**Comment by JMare** [comment]
@vooon I will make the changes you suggested

@TSC21 I will fix the merge errors and do the documentation when I can.

Thanks

@JamesStewy can you please make the recommended changes to the pythons scripts

**Comment by JMare** [comment]
@vooon We have applied all your suggestions!

@TSC21 I have rebased against master to fix merge conflict.

One other thought we had while we are modifying this plugin is to make the current waypoint number more accessible. It seems the only way to access the current waypoint is to iterate through the waypoints array from /mavros/mission/waypoints checking is_current. This is quite clumsy.
What do you think of adding the current seq number outside of the waypoint array in the published msg as we did here?

https://github.com/MonashUAS/mavros/blob/e57ceca1a8ac86501f4a1a31877609229814ee6c/mavros_msgs/msg/WaypointList.msg

**Comment by TSC21** [comment]
I agree with that idea. Can you also please also update the Doxygen documentation in the plugin, similar to what is done in `imu.cpp`?

**Comment by JMare** [comment]
@TSC21 Please take a look at the last commit and see if that is what you are looking for.

I have not used doxygen before so not sure if it is too redundant.

File: mavros/src/plugins/waypoint.cpp [hunk 13]
```diff
@@ -232,7 +232,12 @@ class WaypointPlugin : public plugin::PluginBase {
 
 	/* -*- rx handlers -*- */
 
-	// WaypointItem is MISSION_ITEM child
+	/**
+	 * @brief handle MISSION_ITEM mavlink msg
+	 * handles and stores mission items when pulling waypoints
+	 * @param msg       Received Mavlink msg
```
Comment (TSC21) [inline mavros/src/plugins/waypoint.cpp]: Please use tab instead of whitespaces. (use the uncrustify tool)

**Comment by TSC21** [comment]
@JMare it is good (despite the ident style). Can you do the same for the `low-level send functions` and the `mid-level helpers`? Thanks

File: mavros/src/plugins/waypoint.cpp (hunk 13 above)
Comment (JMare) [reply mavros/src/plugins/waypoint.cpp]: I did use uncrustify but i guess it doesn't fix tabs that aren't at the start of a line.
I've fixed this

File: mavros/src/plugins/waypoint.cpp [hunk 14]
```diff
@@ -513,10 +509,16 @@ class WaypointPlugin : public plugin::PluginBase {
...
 	{
 		lock_guard lock(mutex);
-		if (connected)
+		if (connected) {
 			shedule_pull(BOOTUP_TIME_DT);
-		else
+
+			if (!wp_nh.getParam("enable_partial_push",enable_partial_push)) {
+				enable_partial_push = m_uas->is_ardupilotmega();
+			}
```
Comment (vooon) [inline mavros/src/plugins/waypoint.cpp]: I just checked NodeHandle, and better to use `wp_nh.hasParam()`: http://docs.ros.org/jade/api/roscpp/html/classros_1_1NodeHandle.html#aadd6a7c41de08f54a913e97f26a3af5e

File: mavros/src/plugins/waypoint.cpp [hunk 15]
```diff
@@ -541,6 +545,7 @@ class WaypointPlugin : public plugin::PluginBase {
 		}
 	}
 
+	// @brief Act on first heartbeat from FCU
 	void connection_cb(bool connected) override
```
Comment (vooon) [inline mavros/src/plugins/waypoint.cpp]: Override functions do not need special documentation (exception if it do some not obvious things).

File: mavros/src/plugins/waypoint.cpp [hunk 16]
```diff
@@ -556,6 +561,7 @@ class WaypointPlugin : public plugin::PluginBase {
 		}
 	}
 
+	// @brief Callback for scheduled waypoint pull
```
Comment (TSC21) [inline mavros/src/plugins/waypoint.cpp]: The comment should on the following format to be accepted on Doxygen:
`//! @brief ....`
or
```
  /** @brief ......
   */
```
The first is better for one line comment though.

File: mavros/src/plugins/waypoint.cpp [hunk 17]
```diff
@@ -620,6 +628,8 @@ class WaypointPlugin : public plugin::PluginBase {
 		}
 	}
 
+	// @brief wait until a waypoint pull is complete
```
Comment (TSC21) [inline mavros/src/plugins/waypoint.cpp]: /**
 * @brief wait until a waypoint pull is complete.
 * Pull happens asynchronously, this function blocks until it is done.
 */

File: mavros/src/plugins/waypoint.cpp [hunk 18]
```diff
@@ -628,6 +638,8 @@ class WaypointPlugin : public plugin::PluginBase {
 		       && !is_timedout;
 	}
 
+	// @brief wait until a waypoint push is complete
```
Comment (TSC21) [inline mavros/src/plugins/waypoint.cpp]:
```c++
/**
 * @brief wait until a waypoint pull is complete.
 * Push happens asynchronously. This function blocks until it is done.
 */
```

File: mavros/src/plugins/waypoint.cpp [hunk 19]
```diff
@@ -573,6 +579,7 @@ class WaypointPlugin : public plugin::PluginBase {
 		mission_request_list();
 	}
 
+	// @brief Send ACK back to FCU after pull
```
Comment (TSC21) [inline mavros/src/plugins/waypoint.cpp]: `//! @brief Send ACK back to FCU after pull`

File: mavros/src/plugins/waypoint.cpp (hunk 15 above)
Comment (TSC21) [reply mavros/src/plugins/waypoint.cpp]: @JMare just like this:
`// Act on first heartbeat from FCU` (no `@brief` and no `//!`, which results in it not being processed on Doxygen)

**Comment by JMare** [comment]
Thanks @TSC21 and @vooon I will implement your suggestions but I am going away this weekend and wont be able to do this for a couple of days.

**Comment by JMare** [comment]
Thanks @TSC21 :)

**Pull request merged**